sormas-backend/src/main/java/de/symeda/sormas/backend/util/RightsAllowedInterceptor.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * SORMAS® - Surveillance Outbreak Response Management & Analysis System
+ * Copyright © 2016-2021 Helmholtz-Zentrum für Infektionsforschung GmbH (HZI)
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ * You should have received a copy of the GNU General Public License
+ * along with this program. If not, see <https://www.gnu.org/licenses/>.
+ */
+
+package de.symeda.sormas.backend.util;
+
+import java.util.HashMap;
+
+import javax.annotation.Resource;
+import javax.annotation.security.PermitAll;
+import javax.ejb.EJB;
+import javax.ejb.EJBAccessException;
+import javax.ejb.SessionContext;
+import javax.interceptor.AroundInvoke;
+import javax.interceptor.InvocationContext;
+
+import de.symeda.sormas.api.i18n.I18nProperties;
+import de.symeda.sormas.api.i18n.Strings;
+import de.symeda.sormas.api.utils.AccessDeniedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.symeda.sormas.api.user.UserRight;
+import de.symeda.sormas.backend.user.CurrentUserService;
+
+public class RightsAllowedInterceptor {
+
+	@Resource
+	private SessionContext sessionContext;
+	@EJB
+	private CurrentUserService currentUserService;
+
+	private static final Logger logger = LoggerFactory.getLogger(RightsAllowedInterceptor.class);
+
+	private final HashMap<String, UserRight> userRightValues = new HashMap();
+
+	@AroundInvoke
+	public Object checkRightsAllowed(InvocationContext context) throws Exception {
+
+		PermitAll permitAll = context.getMethod().getAnnotation(PermitAll.class);
+		if (permitAll == null) {
```

Comment:
I don't like this deep if nesting, but `@PermitAll` to be handled as prechecked condition.

```java
// if @PermitAll than don't check UserRights
Optional.ofNullable(context.getMethod().getAnnotation(PermitAll.class)).orElseGet(context.getTarget().getClass().getAnnotation(PermitAll.class)).orElse(null);
if (permitAll != null) {
	return context.proceed();
}
```

Reply:
Being a custom implementation I think for @PermitAll would be nice to check that at least there is a current user so we don't allow access without an authenticated user unless we don't have public features.

Or we could introduce our own @PermitAll like annotation with user restriction and use that one instead of the default one.

Reply:
Rethinking this: The current implementation gives priority of `@PermitAll` on class without checking `@RightsAllowed` on method. That not how Jakarta EE works.

Reply:
> Being a custom implementation I think for @PermitAll would be nice to check that at least there is a current user so we don't allow access without an authenticated user unless we don't have public features.

I disagree: We also need `@PermitAll` for SYSTEM itself (so no user at all).

Reply:
> I disagree: We also need `@PermitAll` for SYSTEM itself (so no user at all).

For SYSTEM we use RightsAllowed(_SYSTEM). That I think will not work for this implementation @MartinWahnschaffe

Reply:
> > I disagree: We also need `@PermitAll` for SYSTEM itself (so no user at all).
> 
> For SYSTEM we use `@RightsAllowed(_SYSTEM)`. That I think will not work for this implementation @MartinWahnschaffe

Yes, it does work. for _SYSTEM the method checks against sessionContext.isCallerInRole